### ISSUE TITLE
Remove the need for remote.CheckPushPermission

### ISSUE THREAD

**Issue by ImJasonH** [google/go-containerregistry#412]
This method employs a somewhat silly hack to determine reliably whether a keychain can be used to authorize pushes to a repo. This can be useful if you want to check whether you can push an image before you spend lots of effort constructing the layers for that image.

@jonjohnsonjr has some ideas about how we can make this unnecessary by allowing users to call `remote.Write` before all the layer contents are available and `mutate.Append`ed in, by passing `stream.Layer`s to it that are populated in goroutines and closed when they're done. This relies on a change to how `remote.Write`, `mutate.Append` and `stream.Layer` fit together.

I'll let him elaborate.

**Comment by jonjohnsonjr** [comment]
I had two ideas for this. One is simple, the other requires a few fixes.

# Simple idea

We create a registry client package that maps directly to the [distribution spec](https://github.com/opencontainers/distribution-spec/blob/master/spec.md). These would be lower-level operations, not like our higher level `remote` package. The `remote` package would use this registry client instead of making http requests directly.

People who need the functionality that `remote.CheckPushPermission` provides could just call `InitiateUpload` directly.

# Complex idea

TL;DR: we need to do everything lazily instead of eagerly. `stream.Layer` is a start down this path.

There are a lot of things we need to fix, that could be broken out into separate issues, but lumping them all in here will make it easier to understand the problem end-to-end, I think.

What needs fixing, in order:
1. `stream.Layer`'s `Size`, `Digest`, and `DiffID` methods should *block* until the stream has been consumed instead of returning `stream.ErrNotComputed`. [`stream.NewLayer`](https://github.com/google/go-containerregistry/blob/8621d738a07bc74b2adeafd175a3c738423577a0/pkg/v1/stream/layer.go#L52) should take the lock and [`compressedReader.Close`](https://github.com/google/go-containerregistry/blob/8621d738a07bc74b2adeafd175a3c738423577a0/pkg/v1/stream/layer.go#L146) should release it. This means that you'll basically always need to use a goroutine to use `stream.Layer` effectively (kind of like io.Pipe), which we should document (or at least provide examples).
2. [`tarball.LayerFromOpener`](https://github.com/google/go-containerregistry/blob/8621d738a07bc74b2adeafd175a3c738423577a0/pkg/v1/tarball/layer.go#L74) should use `stream.Layer` as most of its implementation. Right now, we compute the diffid and digest as soon as we read it, which we should be doing lazily. (For already-compressed layers, we will have to do something else -- computing the DiffID is a pain for those.)
3. [`mutate.image.Layers`](https://github.com/google/go-containerregistry/blob/019cdfc6adf96a4905a1b93a7aeaea1e50c0b6cf/pkg/v1/mutate/mutate.go#L208) should first look through its `Addendum`s to see if there are any streaming layers instead of relying on `compute()` to return `stream.ErrNotComputed`.
4. [`remote.Write`](https://github.com/google/go-containerregistry/blob/8621d738a07bc74b2adeafd175a3c738423577a0/pkg/v1/remote/write.go#L84-L101) should *always* upload the config file in a separate goroutine concurrently with the layers. Because we made `stream.Layer`'s `Size`, `Digest`, and `DiffID` block instead of returning an error, the config upload will sit patiently in a goroutine waiting on any `stream.Layer`s to complete their uploads before continuing. In the case where the image doesn't have any `stream.Layer`s, it will just upload immediately. This simplifies the `remote.Write` code a lot.

## example

Let's look at a code snippet from kaniko to see what's wrong and how this would be better: https://github.com/GoogleContainerTools/kaniko/blob/246cc92a33c29d76b811f7da605375d8bc4c2198/pkg/executor/push.go#L110-L144

### today

We open up a layer using `tarball.LayerFromFile`. Currently, this computes the [Digest, Size, and DiffID immediately](https://github.com/google/go-containerregistry/blob/8621d738a07bc74b2adeafd175a3c738423577a0/pkg/v1/tarball/layer.go#L86-L95).

After that, we initate a push, which can fail if we don't have valid credentials.

### tomorrow

When we open up the layer using `tarball.LayerFromFile`, it will use a `stream.Layer` instead, so the Digest/Size/DiffID computations will be deferred until the layer is consumed.

We'll immediately start the image upload, which will fail when we try to [initiate a blob upload](https://github.com/google/go-containerregistry/blob/8621d738a07bc74b2adeafd175a3c738423577a0/pkg/v1/remote/write.go#L318), which happens before we even try to [open the layer's `io.ReadCloser`](https://github.com/google/go-containerregistry/blob/8621d738a07bc74b2adeafd175a3c738423577a0/pkg/v1/remote/write.go#L330).

Thus, we never even try to open the file.

## followup

Most of the [`mutate`](https://github.com/google/go-containerregistry/blob/master/pkg/v1/mutate/mutate.go) operations are unfortunately synchronous. E.g. they update the image immediately and don't wait lazily for it to be consumed. It's possible that we could introduce a conceptual "thunk" in here such that all these operations are deferred until the corresponding method is actually called again, i.e. we need to make all of those work with `stream.Layer`, though it may be impossible.

I'm not sure if we can do this without breaking clients, but I think it's worth doing even if it's hard.

**Comment by JLLeitschuh** [comment]
Hey Jason,
I'm trying to fix a security vulnerability. This method is no longer used besides in tests. Can I delete it?

**Comment by imjasonh** [comment]
> Hey Jason, I'm trying to fix a security vulnerability. This method is no longer used besides in tests. Can I delete it?

It's an exported method, are we sure it's not used outside of this package?

I'm fine removing it, I don't know of any load-bearing uses of it personally.

edit: there are some users: https://github.com/search?q=remote.CheckPushPermission&type=code
